File: 3_longest_unique_substring/Solution.java
import java.util.HashSet;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int left = 0;
        int right = 0;
        int max = 0;
        
        HashSet<Character> hash_set = new HashSet<>();

        while (right < s.length()) {
            if (!hash_set.contains(s.charAt(right))) {
                hash_set.add(s.charAt(right));
                right++;
                max = Math.max(hash_set.size(), max);
            }
            else {
                hash_set.remove(s.charAt(left));
                left++;
            }
        }
        return max;
    }
}

File: 3_longest_unique_substring/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left, right, longest = 0, 0, 0
        letters = []
        
        while right < len(s):
            char = s[right]
            
            if char not in letters:
                letters.append(char)
                right += 1
                longest = max(len(letters), longest)
            else:
                letters.pop(left)
                
        return longest
        